cmd/soroban-rpc/lib/preflight/src/lib.rs

```diff
@@ -0,0 +1,232 @@
+use std::rc::Rc;
+use xdr::LedgerFootprint;
+
+// TODO: we may want to pass callbacks instead of using global functions
+extern "C" {
+    // LedgerKey XDR in base64 string to LedgerEntry XDR in base64 string
+    fn SnapshotSourceGet(
+        handle: libc::uintptr_t,
+        ledger_key: *const libc::c_char,
+    ) -> *const libc::c_char;
+    // LedgerKey XDR in base64 string to bool
+    fn SnapshotSourceHas(handle: libc::uintptr_t, ledger_key: *const libc::c_char) -> libc::c_int;
+    // Free Strings returned from Go functions
+    fn FreeGoCString(str: *const libc::c_char);
+}
+
+struct CSnapshotSource {
+    handle: libc::uintptr_t,
+}
+
+impl SnapshotSource for CSnapshotSource {
+    fn get(&self, key: &LedgerKey) -> Result<LedgerEntry, HostError> {
+        let key_xdr = key.to_xdr_base64().unwrap();
+        let key_cstr = CString::new(key_xdr).unwrap();
+        let res = unsafe { SnapshotSourceGet(self.handle, key_cstr.as_ptr()) };
+        if res.is_null() {
+            return Err(HostError::from(
+                ScHostStorageErrorCode::AccessToUnknownEntry,
+            ));
+        }
+        let res_cstr = unsafe { CStr::from_ptr(res) };
+        let res_str = res_cstr.to_str().unwrap();
+        // TODO: use a proper error
+        let entry = LedgerEntry::from_xdr_base64(res_str)
+            .map_err(|_| ScHostStorageErrorCode::UnknownError)?;
+        unsafe { FreeGoCString(res) };
+        Ok(entry)
+    }
+
+    fn has(&self, key: &LedgerKey) -> Result<bool, HostError> {
+        let key_xdr = key.to_xdr_base64().unwrap();
+        let key_cstr = CString::new(key_xdr).unwrap();
+        let res = unsafe { SnapshotSourceHas(self.handle, key_cstr.as_ptr()) };
+        Ok(match res {
+            0 => false,
+            _ => true,
+        })
+    }
+}
+
+#[repr(C)]
+pub struct CLedgerInfo {
+    pub protocol_version: u32,
+    pub sequence_number: u32,
+    pub timestamp: u64,
+    pub network_passphrase: *const libc::c_char,
+    pub base_reserve: u32,
+}
+
+impl From<CLedgerInfo> for LedgerInfo {
+    fn from(c: CLedgerInfo) -> Self {
+        let network_passphrase_cstr = unsafe { CStr::from_ptr(c.network_passphrase) };
+        Self {
+            protocol_version: c.protocol_version,
+            sequence_number: c.sequence_number,
+            timestamp: c.timestamp,
+            network_passphrase: network_passphrase_cstr
+                .to_str()
+                .unwrap()
+                .as_bytes()
+                .to_vec(),
+            base_reserve: c.base_reserve,
+        }
+    }
+}
+
+fn storage_footprint_to_ledger_footprint(
+    foot: &storage::Footprint,
+) -> Result<LedgerFootprint, xdr::Error> {
+    let mut read_only: Vec<LedgerKey> = Vec::new();
+    let mut read_write: Vec<LedgerKey> = Vec::new();
+    for (k, v) in &foot.0 {
+        match v {
+            AccessType::ReadOnly => read_only.push((**k).clone()),
+            AccessType::ReadWrite => read_write.push((**k).clone()),
```

Comment:
💡 We don't need to change this now, but fyi I think you can skip cloning the footprint in this conversion. If you change the `foot` parameter to be a move instead of a borrow (i.e. remove the `&`) you should be able to destruct its contents. Moving the footprint should be fine because the `storage` value in the `preflight_host_function_or_maybe_panic` function is only used for this purpose.

This might be a rabbit hole though because of the Rc<>'s.

🤔 @graydon It's pretty inconvenient that the Storage footprint map returned from `Host::try_finish` contains Rc<>'s, especially since users usually interact with it after calling `try_finish` and deconstructing the host, implying that the data is being completely handed over to the user. Maybe there's nothing good we can do about that in the env crate.

Reply:
> you should be able to destruct its contents

how, sorry, but Rc's are super confusing

Reply:
Like this?

```rust
fn storage_footprint_to_ledger_footprint(
    foot: storage::Footprint,
) -> Result<LedgerFootprint, xdr::Error> {
    let mut read_only: Vec<LedgerKey> = Vec::new();
    let mut read_write: Vec<LedgerKey> = Vec::new();
    for (k, v) in foot.0 {
        match v {
            AccessType::ReadOnly => read_only.push(Rc::try_unwrap(k).unwrap()),
            AccessType::ReadWrite => read_write.push(Rc::try_unwrap(k).unwrap()),
        }
    }
    Ok(LedgerFootprint {
        read_only: read_only.try_into()?,
        read_write: read_write.try_into()?,
    })
}
```

Reply:
@leighmcculloch the `Rc<>`s here may be multiply owned. They're to enable the use of `MeteredMap` which treats its elements as shallow-clonable ("like" `RawVal` but, in this case, not `RawVal`) rather than deep-clonable. Recall that `MeteredMap` at this point is an immutable type that duplicates itself entirely whenever modified, and that this is at least partly necessary for both gas metering and transactional semantics when aborting a function -- we should roll back to a previous storage map. So copying storage maps should be "cheap" ish -- O(N) pointer-copies and Rc bumps -- rather than deep copies. Thus the `Rc<>` values are often likely to be multiple references to truly shared data, and a `try_unwrap` is might work but might fail, in which case you need to clone-out.